Create adr-050-improved-trusted-peering.md (#4072)
tendermint/tendermint: docs/architecture/adr-050-improved-trusted-peering.md

# ADR 50: Improved Trusted Peering

## Changelog
* 22-10-2019: Initial draft

## Context

When `max_num_inbound_peers` or `max_num_outbound_peers` of a node reached, the node cannot spare more slots to any peer 
by inbound or outbound. Therefore, after certain period of disconnection, any important peering can be lost indefinitely 
because all slots consumed by other peers, and the node stop trying to dial the peer anymore.

This is happening because of two reasons, exponential backoff and absence of unconditional peering feature for trusted peers.

## Decision

We would like to suggest solving this problem by introducing two parameters in `config.toml`, `wildcard_peer_ids` and 
`maximum_dial_period`. 

> 1) `wildcard_peer_ids`

Reviewer: What about `private_peers_ids`? If we force the node to always connect to private peers, would that be sufficient?

Author: Private and wildcard are having completely different definition. Private means not want to gossip, wildcard means allow connection anytime regardless of max connection.

Putting wildcard into private causes mix of two definitions.

Reviewer 2: I'm not sure wildcard is the best name here, either, since it often implies needing to match some pattern. I like "unconditional"--a word you used yourself a few paragraphs earlier!

Author: unconditional it is! thanks for advice!

A node operator inputs list of ids of peers which are allowed to be connected by both inbound or outbound regardless of 
`max_num_inbound_peers` or `max_num_outbound_peers` of user's node reached or not.

> 2) `maximum_dial_period`

Reviewer: Shouldn't it be called `persistent_peers_max_dial_period` if it's only applicable to persistent peers?

Author: Agree. We will fix the name. Thanks!

Author: Done

> Terms between each dial to each persistent peer will not exceed `maximum_dial_period` during exponential backoff. 

Reviewer: Maybe it's enough to extend the max from 24h to 48h? and lower the exponent?

Author: For validators, monitoring all connection everyday is not very trivial operation. From our experience, running tens of networks sometimes result in missing to alert relatively crucial connection for days. 

For persistent peers, we feel backoff until 30minute is more than enough, from our experience. Never want it to back off 24 or 48 hours.

# ADR 50: Improved Trusted Peering

## Changelog
* 22-10-2019: Initial draft
* 05-11-2019: Modify `maximum_dial_period` to `persistent_peers_max_dial_period`

## Context

> When `max_num_inbound_peers` or `max_num_outbound_peers` of a node reached, the node cannot spare more slots to any peer 

Reviewer 2: ```suggestion
When `max_num_inbound_peers` or `max_num_outbound_peers` of a node is reached, the node cannot spare more slots to any peer 
```

> by inbound or outbound. Therefore, after certain period of disconnection, any important peering can be lost indefinitely 

Reviewer 2: ```suggestion
by inbound or outbound. Therefore, after a certain period of disconnection, any important peering can be lost indefinitely 
```

> because all slots consumed by other peers, and the node stop trying to dial the peer anymore.

Reviewer 2: ```suggestion
because all slots are consumed by other peers, and the node stops trying to dial the peer anymore.
```

This is happening because of two reasons, exponential backoff and absence of unconditional peering feature for trusted peers.

## Decision

We would like to suggest solving this problem by introducing two parameters in `config.toml`, `wildcard_peer_ids` and 
`persistent_peers_max_dial_period`. 

1) `wildcard_peer_ids`

A node operator inputs list of ids of peers which are allowed to be connected by both inbound or outbound regardless of 

> `max_num_inbound_peers` or `max_num_outbound_peers` of user's node reached or not.

Reviewer 2: Are there limits on the number of trusted/wildcard peers a node can specify? And does adding a trusted/wildcard peer effectively increase the maximum number of peers by 1 (e.g. is the maximum number of inbound peers specified by `max_num_inbound_peers` + n, where n is the number of wildcard peers)?

Author: 1. There is no limits for wildcard peers. I think it is OK without the limit. Do we need it?

2. The peers in wildcard_peer_ids is not objective to `max_num_inbound_peers`. So we don't add `n`. The state machine just does not check `max_num_inbound_peers` criteria for any wildcard_peer_ids. Also the state machine just does not count any peer in wildcard_peer_ids to compare `max_num_inbound_peers`

Reviewer: NOTE: we will have a problem with "wildcard_peer_ids is not objective to max_num_inbound_peers" because there's usually a limited number of connections per OS process (`ulimit`)

Reviewer 2: It sounds like we should have some kind of limit here.

Author: I think the OS limit is much far larger than what we need to connect practically. 200 peers are what we practically think maximum, and it is nowhere close to default limit on OS.

Also users can aware that maxpeernum + numofunconditionalpeers = actual possible number of peers. So it is not a unpredictable risk to manage.

For such reason, I think it can be stated in ADR, but we dont think it is needed to be implemented in this spec.

2) `persistent_peers_max_dial_period`

Terms between each dial to each persistent peer will not exceed `persistent_peers_max_dial_period` during exponential backoff. 
Therefore, `dial_period` = min(`persistent_peers_max_dial_period`, exponential_backoff_dial_period)

Alternative approach

Persistent_peers is only for outbound, therefore it is not enough to cover the full utility of `wildcard_peer_ids`. 
@creamers158(https://github.com/Creamers158) suggested putting id-only items into persistent_peers to be handled as 
`wildcard_peer_ids`, but it needs very complicated struct exception for different structure of items in persistent_peers.
Therefore we decided to have `wildcard_peer_ids` to independently cover this use-case.

## Status

Proposed

## Consequences

### Positive

> A node operator can configure two new parameters in `config.toml` so that he/she can assure that tendermint will allow connection

Reviewer 2: ```suggestion
A node operator can configure two new parameters in `config.toml` so that he/she can assure that tendermint will allow connections
```

from/to peers in `wildcard_peer_ids`. Also he/she can assure that every persistent peer will be dialed at least once in every 
`persistent_peers_max_dial_period` term. It achieves more stable and persistent peering for trusted peers.

> ### Negative

Reviewer 2: Are there any possible performance downsides?

Author: Additional calculation what we need for the new features
1) checking whether given peer is listed in `wildcard_peer_ids`
2) for `persistent_peers`, the state machine will dial indefinitely with given period

We think both calculation is not a significant factor for performance downsides.

The new feature introduces two new parameters in `config.toml` which needs explanation for node operators.

### Neutral

## References

> * two p2p feature inhancement proposal(https://github.com/tendermint/tendermint/issues/4053)

Reviewer 2: Can you clean up the whitespace here and also fix the spelling (no worries, but it should be "enhancement," thank you!).

Reviewer: Fixed.
